pkg/fanal/cache/redis.go

```diff
@@ -131,11 +131,8 @@ func (c RedisCache) Close() error {
 func (c RedisCache) Clear() error {
 	ctx := context.Background()
 
-	var cursor uint64
 	for {
-		var keys []string
-		var err error
-		keys, cursor, err = c.client.Scan(ctx, cursor, redisPrefix+"::*", 100).Result()
+		keys, cursor, err := c.client.Scan(ctx, 0, redisPrefix+"::*", 100).Result()
```

Comment:
`github.com/alicebob/miniredis/v2` added support of `count` for `Scan` function - https://github.com/alicebob/miniredis/pull/346

I updated `Clear()` function, because this function was not working correctly before and we couldn't check it using `miniredis`:
We don't need to use `cursor` in `Scan` function, because we clear received keys.
Old mistaken logic:
1) take [0:100] keys. len(keys) == 200
2) clear [0:100] keys. len(keys) == 100.
3) take [100:200] keys, but len(keys) == 100, because we cleared first 100 keys.

Reply:
@knqyf263 Can you take a look at these changes?
We may also want to move this change to another PR or rename this PR so that we don't miss this change and it will be easier to find this change.

Reply:
>1. take [0:100] keys. len(keys) == 200

Is this because miniredis didn't support `count` and returned all keys?

>3. take [100:200] keys, but len(keys) == 100, because we cleared first 100 keys.

Do you mean `len(keys) == 0`? `len(keys) == 100` looks correct.

>- ERR wrong number of arguments for 'unlink' command

Also, why did this error happen? I thought the key length was zero, which caused the error.

Reply:
>We may also want to move this change to another PR or rename this PR so that we don't miss this change and it will be easier to find this change.

Yes, another PR seems better. Thanks for your suggestion. BTW, do we need `cursor`?

```
if len(keys) == 0 {
	break
}
```

doesn't work?

Reply:
> Is this because miniredis didn't support count and returned all keys?

Right.

> Do you mean len(keys) == 0? len(keys) == 100 looks correct.

`TestRedisCache_Clear` test adds 200 keys - https://github.com/aquasecurity/trivy/blob/ec02a655affb9d79f6586e4f16195771fd123c24/pkg/fanal/cache/redis_test.go#L489-L491

> Also, why did this error happen? I thought the key length was zero, which caused the error.

for [100:200] keys (after clearing the first 100 keys) we get `keys` == 0 (because after first clearing len(keys) == 100 and `cursor` == 100).
`miniredis` returns error if you run `UNLINK` function without arguments - https://github.com/alicebob/miniredis/blob/3a21035691c0b46ace87931aa4108a008bcfc384/cmd_generic.go#L291-L295

Reply:
> Yes, another PR seems better

I will open new PR.

> BTW, do we need cursor?

There seems to be no difference.
We can use `cursor` or `len(keys)`.

Reply:
>for [100:200] keys (after clearing the first 100 keys) we get keys == 0 (because after first clearing len(keys) == 100 and cursor == 100).

Yes, that's what I expected.

>take [100:200] keys, but len(keys) == 100, because we cleared first 100 keys.

That was your mistake, right?

`take [100:200] keys, but len(keys) == 0, because we cleared first 100 keys.` makes sense. Not 100, but 0. Am I still missing something?

Reply:
I recheck:
`cursor` is better way.
We don't need to run `c.client.Scan` again to get `keys` == 0.
for our case:
1 run - get first 100 keys ( keys == 100, cursor == 100)
2 run - get second 100 keys (keys == 100, cursor == 0) - break if you cursor
3 run - get empty keys - break if use len(keys)

Reply:
>2 run - get second 100 keys (keys == 100, cursor == 0) - break if you cursor

Why will the cursor be 0, not 100?

Reply:
> That was your mistake, right?

I needed to separate `keys` received from `Scan` function and `keys` stored in Redis 😄 
We were both right.

our test before changes:
1) we save 200 keys in Redis
2) we get (Scan function) 100 keys from Redis (keys == 100, redisKeys == 200, cursor == 100)
3) we clear 100 keys obtained from step 2 (redisKeys == 100, cursor == 100)
4) (new iteration of loop) we try to get second 100 keys, but we use cursor == 100, so we try to get [100:200] keys, but redisKeys == 100, so `Scan` function return keys == 0.

Reply:
> Why will the cursor be 0, not 100?

Because we got all keys- https://github.com/alicebob/miniredis/blob/3a21035691c0b46ace87931aa4108a008bcfc384/cmd_generic.go#L652-L654

Reply:
> I will open new PR.

Created #6056

Reply:
>I needed to separate keys received from Scan function and keys stored in Redis 😄

Oh, I see. Since we were talking about the returned keys from `miniredis`, I thought it meant the length of returned keys. Thanks.

Reply:
>Because we got latest keys - https://github.com/alicebob/miniredis/blob/3a21035691c0b46ace87931aa4108a008bcfc384/cmd_generic.go#L652-L654

Great catch. Using `cursor` is smarter, then.